src/core/coap/coap_message.cpp

```diff
@@ -207,6 +228,154 @@ otError Message::AppendUriQueryOption(const char *aUriQuery)
     return AppendStringOption(OT_COAP_OPTION_URI_QUERY, aUriQuery);
 }
 
+#if OPENTHREAD_CONFIG_COAP_BLOCKWISE_TRANSFER_ENABLE
+otError Message::ReadBlockOptionValues(otCoapOptionType aBlockType)
+{
+    otError             error                     = OT_ERROR_NONE;
+    uint8_t             buf[kMaxOptionHeaderSize] = {0};
+    const otCoapOption *option                    = NULL;
+    OptionIterator      iterator;
+
+    VerifyOrExit((aBlockType == OT_COAP_OPTION_BLOCK1) || (aBlockType == OT_COAP_OPTION_BLOCK2),
+                 error = OT_ERROR_INVALID_ARGS);
+
+    SuccessOrExit(error = iterator.Init(this));
+    VerifyOrExit((option = iterator.GetOptionByNumber(aBlockType)) != NULL, error = OT_ERROR_NOT_FOUND);
+    SuccessOrExit(error = iterator.GetOptionValue(buf));
+
+    SetBlockWiseBlockNumber(0);
+    SetMoreBlocksFlag(false);
+
+    switch (option->mLength)
+    {
+    case 1:
+        SetBlockWiseBlockNumber((buf[0] & 0xf0) >> 4);
+        if ((buf[0] & 0x08) >> 3 == 1)
+        {
+            SetMoreBlocksFlag(true);
+        }
+        SetBlockWiseBlockSize((otCoapOptionBlockSize)(buf[0] & 0x07));
+        break;
+    case 2:
+        SetBlockWiseBlockNumber((buf[0] << 4) + ((buf[1] & 0xf0) >> 4));
+        if ((buf[1] & 0x08) >> 3 == 1)
+        {
+            SetMoreBlocksFlag(true);
+        }
+        SetBlockWiseBlockSize((otCoapOptionBlockSize)(buf[1] & 0x07));
+        break;
+    case 3:
+        SetBlockWiseBlockNumber((buf[0] << 12) + (buf[1] << 4) + ((buf[2] & 0xf0) >> 4));
+        if ((buf[2] & 0x08) >> 3 == 1)
+        {
+            SetMoreBlocksFlag(true);
+        }
+        SetBlockWiseBlockSize((otCoapOptionBlockSize)(buf[2] & 0x07));
+        break;
+    default:
+        error = OT_ERROR_INVALID_ARGS;
+        break;
+    }
+
+exit:
+    return error;
+}
+#endif // OPENTHREAD_CONFIG_COAP_BLOCKWISE_TRANSFER_ENABLE
+
+otError Message::ReadBlockOptionValues(otCoapOptionType       aBlockType,
+                                       uint32_t *             aBlockNumber,
+                                       bool *                 aMoreBlocks,
+                                       otCoapOptionBlockSize *aBlockSize)
+{
+    otError             error                     = OT_ERROR_NONE;
+    uint8_t             buf[kMaxOptionHeaderSize] = {0};
+    const otCoapOption *option                    = NULL;
+    OptionIterator      iterator;
+
+    VerifyOrExit((aBlockType == OT_COAP_OPTION_BLOCK1) || (aBlockType == OT_COAP_OPTION_BLOCK2),
+                 error = OT_ERROR_INVALID_ARGS);
+
+    SuccessOrExit(error = iterator.Init(this));
+    VerifyOrExit((option = iterator.GetOptionByNumber(aBlockType)) != NULL, error = OT_ERROR_NOT_FOUND);
+    SuccessOrExit(error = iterator.GetOptionValue(buf));
+
+    *aBlockNumber = 0;
+    *aMoreBlocks  = false;
+
+    switch (option->mLength)
+    {
+    case 1:
+        *aBlockNumber = (buf[0] & 0xf0) >> 4;
+        if ((buf[0] & 0x08) >> 3 == 1)
```

Comment:
I mentioned about "magic numbers"…

`0x08`, `3`… there's a few numbers on this line that have no apparent meaning without reading the specs.  (_I_ know `0x08` here is the `M` bit, but a casual reader may not!).  PR #4396 adds a routine that will do the donkey work of converting that option to an unsigned integer so you'd be left with:

```
    otError Message::ReadBlockOptionValues(…)
    {
        uint64_t rawValue;
        OptionIterator iterator;
        // …

        SuccessOrExit(error = iterator.Init(this));
        VerifyOrExit((option = iterator.GetFirstOptionMatching(aBlockType)) != NULL, error = OT_ERROR_NOT_FOUND);
        SuccessOrExit(error = iterator.GetOptionValue(rawValue));

        *aBlockNumber = rawValue >> 4;
        if ((rawValue & 0x08) >> 3 == 1)
        {
            *aMoreBlocks = true;
        }
        *aBlockSize = (otCoapOptionBlockSize)(rawValue & 0x07); // and this should probably be a static_cast<>
    }
```

We cut some repetition, but note the magic numbers still there.

As a counter-example, have a look at this:
https://github.com/vrtsystems/openthread/commit/7f0e7f1b638d3faa7cde49425ebed2f891160a92

There, I've put the magic numbers in a macro which is named.  Arguably, I could do a bit more.  Those last few lines would collapse to this:

```
        *aBlockNumber = OT_COAP_BLOCK_NUM(rawValue);
        *aMoreBlocks = OT_COAP_BLOCK_MORE(rawValue) != 0;
        *aBlockSize = (otCoapOptionBlockSize)(OT_COAP_BLOCK_SZX(rawValue));
```